src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/OrderAddItemHintReceiver.cs

```diff
@@ -0,0 +1,103 @@
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Evaluation;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands.Ordering
+{
+    [Export(typeof(IProjectChangeHintReceiver))]
+    [Export(typeof(OrderAddItemHintReceiver))]
+    [ProjectChangeHintKind(ProjectChangeFileSystemEntityHint.AddedFileAsString)]
+    [AppliesTo(ProjectCapability.SortByDisplayOrder)]
+    internal class OrderAddItemHintReceiver : IProjectChangeHintReceiver
+    {
+        private readonly IProjectAccessor _accessor;
+
+        private ImmutableHashSet<string> _previousIncludes = ImmutableHashSet<string>.Empty;
+        private OrderAddItemHintReceiverAction _action = OrderAddItemHintReceiverAction.NoOp;
+        private IProjectTree _target = null;
+
+        [ImportingConstructor]
+        public OrderAddItemHintReceiver(IProjectAccessor accessor)
+        {
+            Requires.NotNull(accessor, nameof(accessor));
+
+            _accessor = accessor;
+        }
+
+        public async Task HintedAsync(IImmutableDictionary<Guid, IImmutableSet<IProjectChangeHint>> hints)
+        {
+            if (_action != OrderAddItemHintReceiverAction.NoOp && _target != null && hints.ContainsKey(ProjectChangeFileSystemEntityHint.AddedFile))
+            {
+                var hintSet = hints[ProjectChangeFileSystemEntityHint.AddedFile];
+                foreach (var hint in hintSet)
+                {
+                    var configuredProject = hint.UnconfiguredProject.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
+
+                    await _accessor.OpenProjectForWriteAsync(configuredProject, project =>
+                    {
+                        // We do a sanity re-evaluation to absolutely ensure changes were met.
+                        project.ReevaluateIfNecessary();
+                        var addedElements = GetAddedElements(_previousIncludes, project);
+
+                        switch (_action)
+                        {
+                            case OrderAddItemHintReceiverAction.MoveToTop:
+                                OrderingHelper.TryMoveElementsToTop(project, addedElements, _target);
+                                break;
+                            case OrderAddItemHintReceiverAction.MoveAbove:
+                                OrderingHelper.TryMoveElementsAbove(project, addedElements, _target);
+                                break;
+                            case OrderAddItemHintReceiverAction.MoveBelow:
+                                OrderingHelper.TryMoveElementsBelow(project, addedElements, _target);
+                                break;
+                            default:
+                                break;
+                        }
+                    }).ConfigureAwait(false);
+                }
+            }
+        }
+
+        public async Task HintingAsync(IProjectChangeHint hint)
```

Comment:
The fact that this takes a single hint, but `HintedAsync` takes a set makes me thing that this might get called many times if several files are added.  Should we ensure that `_previousIncludes` is only set once?

Reply:
Makes sense. I can add logic to make it only set once when the action is not a NoOp.